MAINT: handle edge case in rfftfreq and fftfreq

HashiramaP opened the pull request:
If n = 0, it would imply that there are no data points to work with, and hence no frequency bins to return. This would make the function's output meaningless.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

HashiramaP committed:
added edge case to rfftfreq and fftfreq

 If n = 0, it would imply that there are no data points to work with, and hence no frequency bins to return. This would make the function's output meaningless.

HashiramaP commented:
if not isinstance(n, integer_types):
        raise ValueError("n should be an integer")
This checks if n is an interger
The problem arises because n can still be a valid integer (e.g., 0 or -1) that is not appropriate for the rfftfreq and fftfreq functions. The functions expect n to be a positive integer because:

Zero (n=0): The Fourier transform requires a non-zero window length to calculate frequency bins. A window length of zero would not make sense mathematically for the FFT, as you can’t compute frequencies from zero data points.

Negative (n<0): A negative window length doesn't make sense in the context of Fourier transforms. It would result in invalid or nonsensical calculations.

    if not isinstance(n, integer_types) or n <= 0:
        raise ValueError("n should be a positive integer")
 ensures that n is a positive integer, avoiding these issues and providing a clear error message for invalid input.

seberg commented on numpy/fft/_helper.py:
```diff
@@ -212,6 +214,11 @@ def rfftfreq(n, d=1.0, device=None):
+    ------
+    ValueError
+        If `n` is not a positive integer.
+
```
I don't care much, but not sure it is worth documenting such clearly wrong error (i.e. nobody should have to try/except it).

seberg commented on numpy/fft/_helper.py:
```diff
@@ -163,10 +167,9 @@ def fftfreq(n, d=1.0, device=None):
     """
-    if not isinstance(n, integer_types):
-        raise ValueError("n should be an integer")
+    if not isinstance(n, integer_types) or n <= 0:
```
These days, it would actually be better to just use:
```
n = operator.index(n)
```
and then check `n <= 0`.  (And that changes the error type to `TypeError` for many things, which I think is completely fine or actually better).

seberg commented on numpy/fft/_helper.py:
```diff
@@ -121,14 +121,13 @@ def ifftshift(x, axes=None):
 
     return roll(x, shift, axes)
 
-
```
I don't understand where this change comes from?  (Shouldn't the linter even dislike it.)

The double space after `.` below doesn't matter, but is not really incorrect, just a style choice.
(Any good text editing program will make the space after a period longer, it depends on where you learned things whether or not you prefer two spaces to indicate this.)

seberg reviewed (commented):
FWIW, an explicit check seems maybe slightly nicer.  (To note for other reviewers:  Currently a `ZeroDivisionError` is raised, so it isn't like you get bad results).

A bit opinionated comments and maybe you like to do the `operator.index` thing just to make this code a bit cleaner.  (`operator.index` would not have existed at the time this was written)

seberg commented:
Also, if you add an explicit error here, please make sure we have a test that an error is raised.  Even if it is obvious.

HashiramaP committed:
Update _helper.py with the changes seberg mentioned without tests

HashiramaP commented:
> Also, if you add an explicit error here, please make sure we have a test that an error is raised. Even if it is obvious.

So I did the changes you mentioned, where can I do the tests you mentioned

mattip commented:
`numpy/fft/tests/test_helper.py` seems like the right place

HashiramaP committed:
Added the invalid test cases

HashiramaP commented:
> `numpy/fft/tests/test_helper.py` seems like the right place

Added the Tests, needs revision

seberg commented on numpy/fft/_helper.py:
```diff
@@ -121,14 +122,13 @@ def ifftshift(x, axes=None):
 
     The returned float array `f` contains the frequency bin centers in cycles
-    per unit of the sample spacing (with zero at the start).  For instance, if
+    per unit of the sample spacing (with zero at the start). For instance, if
```
In general, please don't fix this type of thing.  This is perfectly correct, just different habit/style.

seberg commented on numpy/fft/_helper.py:
```diff
@@ -163,19 +163,24 @@ def fftfreq(n, d=1.0, device=None):
+        raise TypeError("n must be an integer")
+    
+    if n <= 0:
+        raise TypeError("n should be a positive integer")
```
This looks like a `ValueError` to me.

seberg commented on numpy/fft/_helper.py:
```diff
@@ -227,9 +232,17 @@ def rfftfreq(n, d=1.0, device=None):
+        raise TypeError("n must be an integer")
+    
+    if n <= 0:
+        raise TypeError("n should be a positive integer")
```
Same, `ValueError` is good.

seberg commented on numpy/fft/_helper.py:
```diff
@@ -163,19 +163,24 @@ def fftfreq(n, d=1.0, device=None):
+    try:
+        n = operator.index(n)
+    except TypeError:
+        raise TypeError("n must be an integer")
```
TBH, I would just keep the original error, but I suppose Ok either way.  The original error gives information about what `n` was, the only downside is that it doesn't mention `n` (but there is only one parameter anyway).

seberg reviewed (commented):
Thanks for the follow up.  I much prefer `operator.index`, and that makes the change more worthwhile to me.

It should use a `ValueError`, though.  And it would be nice if you could revert the whitespace changes.

HashiramaP committed:
changed TypeError to ValueError as seberg said

HashiramaP committed:
changed invalid test cases

They were expecting TypeError, they are now expecting ValueError as seberg suggested

HashiramaP commented:
Hey, was wondering if there was any update?
